src/secret_sharing/malicious_replicated.rs

```diff
@@ -140,38 +146,42 @@ impl<F: Field> Mul<F> for MaliciousReplicated<F> {
     }
 }
 
+#[async_trait]
 impl<F: Field> Downgrade for MaliciousReplicated<F> {
     type Target = Replicated<F>;
-    fn downgrade(self) -> UnauthorizedDowngradeWrapper<Self::Target> {
+    async fn downgrade(self) -> UnauthorizedDowngradeWrapper<Self::Target> {
         UnauthorizedDowngradeWrapper(self.x)
     }
 }
 
+#[async_trait]
 impl<T, U> Downgrade for (T, U)
 where
-    T: Downgrade,
-    U: Downgrade,
+    T: Downgrade + Send,
+    U: Downgrade + Send,
 {
-    type Target = (<T as Downgrade>::Target, <U as Downgrade>::Target);
-    fn downgrade(self) -> UnauthorizedDowngradeWrapper<Self::Target> {
-        UnauthorizedDowngradeWrapper((
-            self.0.downgrade().access_without_downgrade(),
-            self.1.downgrade().access_without_downgrade(),
-        ))
+    type Target = (<T>::Target, <U>::Target);
+    async fn downgrade(self) -> UnauthorizedDowngradeWrapper<Self::Target> {
+        let one = self.0.downgrade();
+        let two = self.1.downgrade();
+        let out = join(one, two).await;
+        let output = (out.0.access_without_downgrade(), out.1.access_without_downgrade());
+        UnauthorizedDowngradeWrapper(output)
     }
 }
 
+#[async_trait]
 impl<T> Downgrade for Vec<T>
 where
-    T: Downgrade,
+    T: Downgrade + Send,
 {
     type Target = Vec<<T as Downgrade>::Target>;
-    fn downgrade(self) -> UnauthorizedDowngradeWrapper<Self::Target> {
-        UnauthorizedDowngradeWrapper(
-            self.into_iter()
-                .map(|v| v.downgrade().access_without_downgrade())
-                .collect(),
-        )
+    async fn downgrade(self) -> UnauthorizedDowngradeWrapper<Self::Target> {
+        let output = join_all(self
+            .into_iter()
+            .map(|v| async move { v.downgrade().await.access_without_downgrade() })
+            .collect::<Vec<_>>()).await;
+        UnauthorizedDowngradeWrapper(output)
```

Comment:
@martinthomson following error is seen:
future cannot be sent between threads safely
the trait `std::marker::Send` is not implemented for `<T as Downgrade>::Target`
required for the cast from `impl futures::Future<Output = UnauthorizedDowngradeWrapper<Vec<<T as Downgrade>::Target>>>` to the object type `dyn futures::Future<Output = UnauthorizedDowngradeWrapper<Vec<<T as Downgrade>::Target>>> + std::marker::Send`rustc[Click for full compiler diagnostic](rust-analyzer-diagnostics-view:/diagnostic%20message?7#file%3A%2F%2F%2FUsers%2Frichaj%2FProposals%2FIPA%2Fraw-ipa%2Fsrc%2Fsecret_sharing%2Fmalicious_replicated.rs)
malicious_replicated.rs(183, 34): future is not `Send` as this value is used across an await
malicious_replicated.rs(183, 40): the value is later dropped here